### PR TITLE
Np 3130 map hrcs codes

### PR DESCRIPTION
**Pull request by Eirik-Eide-Saur**

### CODE REVIEW

File: cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinHrcsCategoriesAndActiveties.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    public static final String HRCS_ACTIVITY_URI = "https://nva.unit.no/hrcs/activity/";
+
+    @JsonProperty(CATEGORY)
+    private String category;
+    @JsonProperty(ACTIVITY)
+    private String activity;
+
+    public static String insertCategoryIdIntoUriString(String categoryId) {
+        //NVA uses 0-indexing while Cristin starts at 1...
+        String categoryIdSubtractedByOne = subtractNumericValueOfStringByOne(categoryId);
```
Comment (axthosarouris) [inline cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinHrcsCategoriesAndActiveties.java:35]: adjustCategoryIdIndex

File: cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinHrcsCategoriesAndActiveties.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    public static final String HRCS_CATEGORY_URI = "https://nva.unit.no/hrcs/category/";
+    public static final String HRCS_ACTIVITY_URI = "https://nva.unit.no/hrcs/activity/";
+
+    @JsonProperty(CATEGORY)
+    private String category;
+    @JsonProperty(ACTIVITY)
+    private String activity;
+
+    public static String insertCategoryIdIntoUriString(String categoryId) {
+        //NVA uses 0-indexing while Cristin starts at 1...
```
Comment (axthosarouris) [inline cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinHrcsCategoriesAndActiveties.java:34]: This should be a comment in the method adjustCategoryIndex

File: cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinHrcsCategoriesAndActiveties.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        String categoryIdSubtractedByOneWithLeadingZeroes = addRequiredNumberOfLeadingZeros(categoryIdSubtractedByOne);
+        return HRCS_CATEGORY_URI + categoryIdSubtractedByOneWithLeadingZeroes;
+    }
+
+    private static String subtractNumericValueOfStringByOne(String originalNumericString) {
+        return String.valueOf(Integer.parseInt(originalNumericString) - 1);
+    }
+
+    private static String addRequiredNumberOfLeadingZeros(String numberAsString) {
+        return numberAsString.length() == 1 ? "00" + numberAsString : "0" + numberAsString;
```
Comment (axthosarouris) [inline cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinHrcsCategoriesAndActiveties.java:45]: https://stackoverflow.com/questions/6034523/format-an-integer-using-java-string-format

File: cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinHrcsCategoriesAndActiveties.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        buildMethodName = "build",
+        setterPrefix = "with"
+)
+@AllArgsConstructor(access = AccessLevel.PACKAGE)
+public class CristinHrcsCategoriesAndActiveties {
+
+    private static final String CATEGORY = "helsekategorikode";
+    private static final String ACTIVITY = "aktivitetskode";
+
+    public static final String HRCS_CATEGORY_URI = "https://nva.unit.no/hrcs/category/";
```
Comment (axthosarouris) [inline cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinHrcsCategoriesAndActiveties.java:25]: This is not going to work for any other environment except for production. We have done similar work in the past, use the env variables for the custom domain.

File: cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinHrcsCategoriesAndActiveties.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    @JsonProperty(CATEGORY)
+    private String category;
+    @JsonProperty(ACTIVITY)
+    private String activity;
+
+    public static String insertCategoryIdIntoUriString(String categoryId) {
+        //NVA uses 0-indexing while Cristin starts at 1...
+        String categoryIdSubtractedByOne = subtractNumericValueOfStringByOne(categoryId);
+        String categoryIdSubtractedByOneWithLeadingZeroes = addRequiredNumberOfLeadingZeros(categoryIdSubtractedByOne);
+        return HRCS_CATEGORY_URI + categoryIdSubtractedByOneWithLeadingZeroes;
```
Comment (axthosarouris) [inline cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinHrcsCategoriesAndActiveties.java:37]: Create the URI here. Why are you returning a String that is a URI and not directly a URI?

File: cristin-import/src/test/resources/features/GeneralMappingRules.feature [hunk 6]
```diff
@@ -211,6 +211,19 @@ Feature: Mappings that hold for all types of Cristin Results
     When the Cristin Result is converted to an NVA Resource
     Then the NVA Resource has no projects
 
+  Scenario: The Cristin Result's HRCS values are used to generate the URIs for the NVA Resource
+    Given a valid Cristin Result
+    And the Cristin Result has the HRCS values:
+      | helsekategorikode | aktivitetskode |
+      | 4                 | 6.4            |
+      | 13                | 3.1            |
```
Comment (axthosarouris) [inline cristin-import/src/test/resources/features/GeneralMappingRules.feature:219]: 3 digits means that there can be a 3 digit category. How about helsekategorikodes with 3 digits?

**Review by axthosarouris — CHANGES_REQUESTED** [review]

File: cristin-import/src/test/resources/features/GeneralMappingRules.feature [hunk 7]
```diff
@@ -211,6 +211,19 @@ Feature: Mappings that hold for all types of Cristin Results
     When the Cristin Result is converted to an NVA Resource
     Then the NVA Resource has no projects
 
+  Scenario: The Cristin Result's HRCS values are used to generate the URIs for the NVA Resource
```
Comment (axthosarouris) [inline cristin-import/src/test/resources/features/GeneralMappingRules.feature:214]: Do we take care of helsekategorikode and aktivitetskode values that have the expected format but they do not exist?

**Review by axthosarouris — CHANGES_REQUESTED** [review]
It seems that not all

File: cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinHrcsCategoriesAndActiveties.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        return originalId - 1;
+    }
+
+    private static String addRequiredNumberOfLeadingZeros(int categoryId) {
+        return String.format("%03d", categoryId);
+    }
+
+
+    @JacocoGenerated
+    public CristinHrcsCategoriesAndActiveties.CristinHrcsBuilder copy() {
```
Comment (brinxmat) [inline cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinHrcsCategoriesAndActiveties.java:93]: s/Activeties/Activities/ in all occurrences

File: cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinMapper.java [hunk 9]
```diff
@@ -157,6 +159,39 @@ private EntityDescription generateEntityDescription() {
             .collect(Collectors.toList());
     }
 
+    private List<URI> generateNvaHrcsCategoriesAndActivities() {
+        if (extractCristinHrcsCategoriesAndActivities() == null) {
```
Comment (brinxmat) [inline cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinMapper.java:163]: isNull(extractCristinHrcsCategoriesAndActivities()), or return an optional from the method called

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinHrcsCategoriesAndActiveties.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        double activityAsInt;
+        try {
+            activityAsInt = Double.parseDouble(activity);
+        } catch (NumberFormatException e) {
+            throw new HrcsException(String.format(CATEGORY_OR_ACTIVITY_NOT_A_NUMBER_EXCEPTION, "activity", activity));
+        }
+        if (activity.length() != ACCEPTED_ACTIVITY_STRING_LENGTH) {
+            throw new HrcsException(String.format(ACTIVITY_WRONG_FORMAT_EXCEPTION, activity));
+        }
+        if (activityAsInt < 1 || activityAsInt > 8.5) {
```
[the remark this answers is not in the document]
Comment (axthosarouris) [reply cristin-import/src/main/java/no/unit/nva/cristin/mapper/CristinHrcsCategoriesAndActiveties.java:76]: 1 and 8.5 should become constants.  But I agree with Rurik, that it makes more sense to treat these values as a small set of valid values that should be probably hardcoded to a Map of CristinValue->NvaValue pairs

**Review by axthosarouris — APPROVED** [review]

**Review by brinxmat — APPROVED** [review]

**Pull request merged**